### PR TITLE
Some random NFT improvements

### PR DESCRIPTION
**Pull request by tomquirk**
- use sortNftsByContributionFloor where possible
-

### CODE REVIEW

File: src/constants/transactionDefaults.ts [hunk 1]
```diff
@@ -2,3 +2,5 @@ export const DEFAULT_MEMO = ''
 export const DEFAULT_METADATA = 0
 export const DEFAULT_MIN_RETURNED_TOKENS = 0
 export const DEFAULT_ALLOW_OVERSPENDING = true
+export const DEFAULT_JB_721_TIER_CATEGORY = 1
```
Comment (tomquirk) [inline src/constants/transactionDefaults.ts:5]: random 3.1 stuff coming soon

File: src/hooks/JB721Delegate/HasNftRewards.ts [hunk 2]
```diff
@@ -22,19 +21,13 @@ export function useHasNftRewards(): { value: boolean; loading: boolean } {
...
     loading: { fundingCycleLoading },
   } = useContext(V2V3ProjectContext)
-  const { value: supportsV1Interface, loading: v1Loading } =
-    useIsJB721DelegateV1({
-      dataSourceAddress: fundingCycleMetadata?.dataSource,
-    })
-  const { value: supportsV1_1Interface, loading: v1_1Loading } =
-    useIsJB721DelegateV1_1({
-      dataSourceAddress: fundingCycleMetadata?.dataSource,
-    })
```
Comment (tomquirk) [inline src/hooks/JB721Delegate/HasNftRewards.ts:32]: yeah so...this was a big issue lol. we were calling out to infura every time this hook is called, which is a lot.

File: src/hooks/JB721Delegate/transactor/AdjustTiersTx.ts [hunk 3]
```diff
@@ -36,14 +36,7 @@ export function useAdjustTiersTx({
       return Promise.resolve(false)
     }
 
-    // Contract expects tiers to be sorted by price.
-    const newTiersAscending = newTiers.sort((a, b) => {
```
Comment (tomquirk) [inline src/hooks/JB721Delegate/transactor/AdjustTiersTx.ts:40]: We've already sorted it in buildJBtierParams

File: src/models/nftRewardTier.ts [hunk 4]
```diff
@@ -43,6 +43,17 @@ export interface JB721TierParams {
   transfersPausable: boolean
 }
 
+export type JB_721_TIER_PARAMS_V1_1 = Omit<
```
Comment (tomquirk) [inline src/models/nftRewardTier.ts:46]: 1.1 shit

File: src/utils/nftRewards.ts [hunk 5]
```diff
@@ -65,6 +57,14 @@ export async function findJBTiered721DelegateStoreAddress() {
   )?.contractAddress
 }
 
+export function sortNftsByContributionFloor(
```
Comment (tomquirk) [inline src/utils/nftRewards.ts:60]: just moved the fn

File: src/utils/nftRewards.ts [hunk 6]
```diff
@@ -225,9 +225,42 @@ export function tiersEqual({
   )
 }
 
-// Builds JB721TierParams[] (see juice-721-delegate:structs/JB721TierParams.sol)
+function nftRewardTierToJB721TierParams(
```
Comment (tomquirk) [inline src/utils/nftRewards.ts:228]: broke this out into own file

File: src/contexts/NftRewards/NftRewardsProvider.tsx [hunk 7]
```diff
@@ -20,8 +19,12 @@ export const NftRewardsProvider: React.FC = ({ children }) => {
   const { projectMetadata, projectId } = useContext(ProjectMetadataContext)
 
   const dataSourceAddress = fundingCycleMetadata?.dataSource
+  const contractVersion = useJB721DelegateVersion({
+    dataSourceAddress,
+  })
+
   // don't fetch stuff if there's no datasource in the first place.
-  const { value: hasNftRewards } = useHasNftRewards()
+  const hasNftRewards = Boolean(contractVersion)
```
Comment (tomquirk) [inline src/contexts/NftRewards/NftRewardsProvider.tsx:27]: Now we only call `useJB721DelegateVersion` once per pageload. hasNftRewards will be `false` if no contractVersion.

File: src/hooks/JB721Delegate/HasNftRewards.ts (hunk 2 above)
Comment (johnnyd-eth) [reply src/hooks/JB721Delegate/HasNftRewards.ts:32]: Gotcha, good catch.

**Review by johnnyd-eth — APPROVED** [review]
Epic PR, loading state is way better. LGTM

File: src/utils/nftRewards.ts [hunk 8]
```diff
@@ -237,36 +270,8 @@ export function buildJB721TierParams({
...
-        reservedTokenBeneficiary,
-        encodedIPFSUri,
-        allowManualMint: false,
-        shouldUseBeneficiaryAsDefault: false,
-        transfersPausable: false,
-      }
+      const rewardTier = _rewardTiers[index]
+      return nftRewardTierToJB721TierParams(rewardTier, cid)
     })
     .sort((a, b) => {
```
Comment (johnnyd-eth) [inline src/utils/nftRewards.ts:276]: **Question:** Wondering if we should have `.slice()` in here for safety, like in `SortNftsByContributionFloor`?

Comment (tomquirk) [reply src/utils/nftRewards.ts:278]: done

**Pull request merged**